File: 3HC_Firmware/3HC_Firmware_0_1.cydsn/NMEA0183.c
/* ========================================
 * FILE:   NMEA0183.c
 * AUTHOR: Caleb Cotter
 *
 * CREATED ON FEBRUARY 24, 2020, 1:25 PM
 *
 * ========================================
 *
 * Copyright SLUGSAT, 2020
 * All Rights Reserved
 * UNPUBLISHED, LICENSED SOFTWARE.
 *
 * CONFIDENTIAL AND PROPRIETARY INFORMATION
 * WHICH IS THE PROPERTY OF SLUGSAT.
 *
 * ========================================
*/

#include "NMEA0183.h"

/******************************************************************************
 * Private Functions                                                          *  
 ******************************************************************************/

uint8_t StringChecksum(char * string) {

    //Local Variables
    int stringLength = strlen(string);
    int i;
    uint8_t checksum = 0x00;

    //Calculate One-Byte XOR Checksum
    for (i = 0; i < stringLength; i++) {
        checksum ^= string[i];
    }
    return checksum;
}

uint8_t AsciiToInt(char letter) {
    switch (letter) {
        case '0'...'9':
            return (letter - 48);
            break;
        case 'a'...'f':
            return (letter - 87);
            break;
        case 'A'...'F':
            return (letter - 55);
            break;
    }
    return -1;
}

/******************************************************************************
 * Public Functions                                                           *  
 ******************************************************************************/

int NMEA0183_DecodeMsg(uint8 * msg, char * id, char * payload){
    // Input Checking
    if(msg == NULL || id == NULL || payload == NULL){
        return -1;
    }
    
    // Local Variables
    uint8_t temp[MAX_MSG_LENGTH];
    uint8_t checksum;

    // Get Message Components
    if(sscanf((char *) msg,"$%[^,],%[^*]*%02X", id, payload, (unsigned int *) &checksum) != 3){
        return -1;
    }
    
    // Validate Checksum
    sprintf((char *) temp,"%s,%s",id,payload);
    if(checksum != StringChecksum((char *)temp)){
        sprintf(id,"ERR");
        return -1;
    }
    
    return 0;
}

int NMEA0183_EncodeMsg(uint8 * msg, char * id, char * payload){
    // Input Checking
    if(msg == NULL || id == NULL || payload == NULL){
        return -1;
    }
    
    // Local Variables
    uint8_t temp[MAX_MSG_LENGTH];
    uint8_t checksum;
    
    // Build Message and Calc Checksum
    sprintf((char *) temp,"%s,%s", id, payload);
    checksum = StringChecksum((char *)temp);
    
    // Add Checksum to Message
    sprintf((char *)msg,"$%s*%02X\n",temp,checksum);
    
    return 0;
}

/* [] END OF FILE */

File: 3HC_UI_Software/NMEA_0183.py
# /* ========================================
#  * FILE:   NMEA_0183.py
#  * AUTHOR: Nick Jannuzzi
#  *
#  * CREATED ON April 13, 2020, 4:10 PM
#  *
#  * ========================================
#  *
#  * Copyright SLUGSAT, 2020
#  * All Rights Reserved
#  * UNPUBLISHED, LICENSED SOFTWARE.
#  *
#  * CONFIDENTIAL AND PROPRIETARY INFORMATION
#  * WHICH IS THE PROPERTY OF SLUGSAT.
#  *
#  * ========================================
# */



import re
#encodes a message and payload using NMEA_0183 GPS protocol
def Encode(msgid, payload):
	a = str(msgid) + ',' + str(payload)
	
	chk = Checksum(a)
	m = '$'+str(a)+'*'+str(hex(chk)[2:])+"\n"
	# print(m+'\n')
	return m

#calculates bytewise XOR checksum of id and payload
def Checksum(string):
	checksum = 0
	j = 0
	for i in string:
		checksum ^= ord(string[j])
		j+=1
	return checksum	


#decodes NMEA_0183 Messages and returns a list containing the separated id, payload and checksum
def Decode(string):
	# print("Before:",string)
	msg = re.split('[$ , * \n]',string)
	#trims some leading and trailing $ and \n 
	msg.pop(0)
	msg.pop(-1)

	#current sensor had weird hex padding, so this will trim anything else that we don't want
	if(len(msg) > 3):
		while(len(msg) > 3):
			msg.pop(-1)	
	# print("AfterTrim:",msg)

	#save ID, payload and checksum into single variables from tuple
	msgid = msg[0]
	payload = msg[1]
	chk = int(msg[-1],16)


	#checksum comparison determines if the message is valid or not
	if(chk != Checksum(str(msgid) + ',' + str(payload))):
		# print("Checksum error, expected checksum did not match received checksum\n")
		msg = ["ERR"]
		return msg	
	return msg




